<a href="https://colab.research.google.com/github/Arunkumarvp/Data-mining-and-machine-learning-/blob/main/DMML_LAB_3_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***24CY731** - **DATA MINING AND MACHINE LERANING IN CYBER SECURITY***

#LAB-3

Arun Kumar P
CB.SC.P2CYS25033

##1) Use StandardScaler to standardize the features of a Credit card fraud dataset Include code, description and screenshots of outputs.

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn.preprocessing
from sklearn.preprocessing import StandardScaler


#import dataset from kaggle
import kagglehub
from kagglehub import KaggleDatasetAdapter
# Set the path to the file you'd like to load

file_path = "creditcard.csv"

# Load the latest version
df = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "mlg-ulb/creditcardfraud",
  file_path,
)

print("Data Sample: " , df.head())

/tmp/ipython-input-4143267421.py:17: DeprecationWarning: Use dataset_load() instead of load_dataset(). load_dataset() will be removed in a future version.
  df = kagglehub.load_dataset(


Using Colab cache for faster access to the 'creditcardfraud' dataset.
Data Sample:     Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]


# Task
Standardize the 'Time' and 'Amount' columns in the `df` DataFrame using `StandardScaler`, replacing the original columns with their scaled versions, then display the first few rows of the modified DataFrame and its descriptive statistics.

## Standardize 'Time' and 'Amount'

### Subtask:
Apply the StandardScaler to the 'Time' and 'Amount' columns of the DataFrame to standardize these features. The original columns will be replaced with their scaled versions.


**Reasoning**:
The subtask requires standardizing the 'Time' and 'Amount' columns using StandardScaler. I will initialize StandardScaler and then apply fit_transform to each column, overwriting the original columns with the scaled values.



In [5]:
scaler = StandardScaler()
df['Time'] = scaler.fit_transform(df[['Time']])
df['Amount'] = scaler.fit_transform(df[['Amount']])
print("DataFrame after standardizing 'Time' and 'Amount' columns:")
print(df[['Time', 'Amount']].head())

DataFrame after standardizing 'Time' and 'Amount' columns:
       Time    Amount
0 -1.996583  0.244964
1 -1.996583 -0.342475
2 -1.996562  1.160686
3 -1.996562  0.140534
4 -1.996541 -0.073403


**Reasoning**:
The 'Time' and 'Amount' columns have been standardized. To confirm this and inspect the modified DataFrame, I will display the first few rows of the entire DataFrame and its descriptive statistics.



In [6]:
print("Modified DataFrame head:")
print(df.head())
print("\nDescriptive statistics of the modified DataFrame:")
print(df.describe())

Modified DataFrame head:
       Time        V1        V2        V3  ...       V27       V28    Amount  Class
0 -1.996583 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  0.244964      0
1 -1.996583  1.191857  0.266151  0.166480  ... -0.008983  0.014724 -0.342475      0
2 -1.996562 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  1.160686      0
3 -1.996562 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  0.140534      0
4 -1.996541 -1.158233  0.877737  1.548718  ...  0.219422  0.215153 -0.073403      0

[5 rows x 31 columns]

Descriptive statistics of the modified DataFrame:
               Time            V1  ...        Amount          Class
count  2.848070e+05  2.848070e+05  ...  2.848070e+05  284807.000000
mean  -3.065637e-16  1.168375e-15  ...  2.913952e-17       0.001727
std    1.000002e+00  1.958696e+00  ...  1.000002e+00       0.041527
min   -1.996583e+00 -5.640751e+01  ... -3.532294e-01       0.000000
25%   -8.552120e-01 -9.203734e-01  ... -3.308401e-01     

## 2) Apply SMOTE to address the class imbalance problem. Include code, description and screenshots of outputs.

## Apply SMOTE for Class Imbalance

### Subtask:
Apply SMOTE (Synthetic Minority Over-sampling Technique) to the dataset to address the class imbalance problem. This involves separating the features and target variable, then using SMOTE to create synthetic samples for the minority class, and finally, examining the new class distribution.

**Reasoning**:
SMOTE (Synthetic Minority Over-sampling Technique) is an oversampling method that generates synthetic samples from the minority class. This helps to balance the class distribution, which is crucial for training machine learning models on imbalanced datasets, as it prevents the model from being biased towards the majority class. I will first import the necessary `SMOTE` class and then prepare the data for application.

In [8]:
from imblearn.over_sampling import SMOTE

# Separate features (X) and target (y)
X = df.drop('Class', axis=1)
y = df['Class']

# Display the original class distribution
print("Original class distribution:")
print(y.value_counts())

# Initialize SMOTE
sm = SMOTE(random_state=42)

# Apply SMOTE to the dataset
X_res, y_res = sm.fit_resample(X, y)

# Display the new class distribution
print("\nNew class distribution after SMOTE:")
print(y_res.value_counts())

Original class distribution:
Class
0    284315
1       492
Name: count, dtype: int64

New class distribution after SMOTE:
Class
0    284315
1    284315
Name: count, dtype: int64


**Reasoning**:
The SMOTE operation has been performed, and the class distribution has been displayed. To provide a clear output and address the user's request for documentation, I will combine the resampled features and target back into a new DataFrame and display its head, along with a description of the class distributions for comparison.

In [10]:
# Combine the resampled features and target into a new DataFrame
df_resampled = pd.concat([X_res, y_res], axis=1)

print("\nDataFrame head after SMOTE:")
print(df_resampled.head())

print("\nDescriptive statistics of the resampled DataFrame:")
print(df_resampled.describe())


DataFrame head after SMOTE:
       Time        V1        V2        V3  ...       V27       V28    Amount  Class
0 -1.996583 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  0.244964      0
1 -1.996583  1.191857  0.266151  0.166480  ... -0.008983  0.014724 -0.342475      0
2 -1.996562 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  1.160686      0
3 -1.996562 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  0.140534      0
4 -1.996541 -1.158233  0.877737  1.548718  ...  0.219422  0.215153 -0.073403      0

[5 rows x 31 columns]

Descriptive statistics of the resampled DataFrame:
                Time             V1  ...         Amount     Class
count  568630.000000  568630.000000  ...  568630.000000  568630.0
mean       -0.159338      -2.315118  ...       0.036056       0.5
std         0.987483       5.473258  ...       0.921006       0.5
min        -1.996583     -56.407510  ...      -0.353229       0.0
25%        -0.987214      -2.750573  ...      -0.342914       

 ## 3) Apply Decision Tree Classifier Model and Evaluate the Performance with Confusion matrix, accuracy, precision, Recall and F1 Score.

# Task
Split the `df_resampled` DataFrame into features (X) and target (y), and then further divide them into training and testing sets using `train_test_split`.

## Prepare Data for Model Training

### Subtask:
Split the resampled dataset (df_resampled) into features (X) and target (y), and then further divide them into training and testing sets using `train_test_split`.


### Reasoning:
Before applying any machine learning model, it's crucial to prepare the data. This involves two main steps:
1. **Separating Features and Target**: The dataset needs to be split into features (input variables, typically denoted as `X`) and the target variable (output variable, typically denoted as `y`). This distinction is fundamental for supervised learning tasks.
2. **Splitting into Training and Testing Sets**: To evaluate the model's performance on unseen data and prevent overfitting, the dataset is further divided into training and testing sets. The training set is used to train the model, while the testing set is used to assess its generalization capability. A `random_state` is set for reproducibility, ensuring that the split is the same every time the code is run.

**Reasoning**:
I will separate the features and target variable from the `df_resampled` DataFrame, then split them into training and testing sets using `train_test_split`, and finally print the shapes of the resulting datasets to verify the split.



In [11]:
from sklearn.model_selection import train_test_split

# Separate features (X) and target (y)
X = df_resampled.drop('Class', axis=1)
y = df_resampled['Class']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Print the shapes of the resulting sets
print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (454904, 30)
Shape of y_train: (454904,)
Shape of X_test: (113726, 30)
Shape of y_test: (113726,)


## Train Decision Tree Classifier

### Subtask:
Initialize and train a `DecisionTreeClassifier` on the training data (X_train, y_train).


**Reasoning**:
I will import the `DecisionTreeClassifier` from `sklearn.tree`, instantiate it with a `random_state` for reproducibility, and then train the model using the `fit` method on the `X_train` and `y_train` datasets.



In [12]:
from sklearn.tree import DecisionTreeClassifier

# Initialize the Decision Tree Classifier
dt_classifier = DecisionTreeClassifier(random_state=42)

# Train the model on the training data
dt_classifier.fit(X_train, y_train)

print("Decision Tree Classifier trained successfully.")

Decision Tree Classifier trained successfully.


**Reasoning**:
The Decision Tree Classifier has been trained. The next step is to use this trained model to make predictions on the test dataset (`X_test`) to prepare for evaluating its performance.



In [13]:
y_pred = dt_classifier.predict(X_test)
print("Predictions made on the test set.")

Predictions made on the test set.


**Reasoning**:
To evaluate the performance of the trained Decision Tree Classifier, I will calculate and display common classification metrics such as the Confusion Matrix, Accuracy, Precision, Recall, and F1-Score. This will provide a comprehensive understanding of the model's performance.



In [14]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

# Calculate evaluation metrics
conf_matrix = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Confusion Matrix:\n", conf_matrix)
print(f"\nAccuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1 Score: {f1:.4f}")

Confusion Matrix:
 [[56610   140]
 [   56 56920]]

Accuracy: 0.9983
Precision: 0.9975
Recall: 0.9990
F1 Score: 0.9983


The Decision Tree Classifier was successfully applied and evaluated. The model demonstrated strong performance in classifying fraudulent transactions.

### Data Analysis Key Findings

*   The dataset was successfully split into training and testing sets, with 80% used for training (454,904 samples) and 20% for testing (113,726 samples).
*   A Decision Tree Classifier was initialized and trained on the resampled training data.
*   The model achieved a high **Accuracy** of 0.9983 on the test set.
*   The **Precision** for the positive class (fraudulent transactions) was 0.9975, indicating a very low rate of false positives.
*   The **Recall** for the positive class was 0.9990, meaning the model correctly identified nearly all actual fraudulent transactions.
*   The **F1 Score** was 0.9983, representing a balanced measure of precision and recall.
*   The confusion matrix shows that out of 113,726 test samples, there were only 140 false positives and 56 false negatives, demonstrating excellent classification performance.

